python/device/device.py

```diff
@@ -758,10 +929,117 @@ def handle_test_c2d_messages_thread(self, worker_thread_info):
                 self.metrics.receive_message_count_received.increment()
                 self.out_of_order_message_tracker.add_message(thief.get("testC2dMessageIndex"))
 
+    def send_control_message_thread(self, worker_thread_info):
+        """
+        Thread which reads the control message queue and sends the message.  We probably don't
+        need this queue since we're not trying to overlap control messages, but it can't hurt
+        and it mirrors how test messages work.
+        """
+        while not self.done.isSet():
+            # We do not check is_pairing_complete here because we use this thread to
+            # send pairing requests.
+            worker_thread_info.watchdog_epochtime = time.time()
+            if self.is_paused():
+                time.sleep(1)
+                continue
+
+            try:
+                msg = self.outgoing_control_message_queue.get(timeout=1)
+            except queue.Empty:
+                msg = None
+            if msg:
+                try:
+                    self.client.send_message(msg)
+                except Exception as e:
+                    logger.warning(
+                        "send_message raised {} sending a control messsage.  Ignoring".format(
```

Comment:
Is there no retry if the message fails to send?

I mean, if the device were disconnected for a bit, would this message be lost?

In the .NET impl, I keep retrying the top message until it goes through.

Reply:
There is a retry, but there could still be a fatal (non-retryable) exception.  Unfortunately, our current Python SDK will hold the message until it sends, with no timeout and no exponential backoff.

Reply:
Maybe a code comment would be helpful, given each SDK has its own behavior in cases like this.